core/inject.js

```diff
@@ -232,71 +232,163 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
   mainWorkspace.translate(0, 0);
   Blockly.mainWorkspace = mainWorkspace;
 
-  if (!options.readOnly && !options.hasScrollbars) {
-    var workspaceChanged = function(e) {
-      if (!mainWorkspace.isDragging()) {
-        var metrics = mainWorkspace.getMetrics();
-        var edgeLeft = metrics.viewLeft + metrics.absoluteLeft;
-        var edgeTop = metrics.viewTop + metrics.absoluteTop;
-        if (metrics.contentTop < edgeTop ||
-            metrics.contentTop + metrics.contentHeight >
-            metrics.viewHeight + edgeTop ||
-            metrics.contentLeft <
-                (options.RTL ? metrics.viewLeft : edgeLeft) ||
-            metrics.contentLeft + metrics.contentWidth > (options.RTL ?
-                metrics.viewWidth : metrics.viewWidth + edgeLeft)) {
-          // One or more blocks may be out of bounds.  Bump them back in.
-          var MARGIN = 25;
-          var blocks = mainWorkspace.getTopBlocks(false);
+  if (!options.readOnly && !mainWorkspace.isMovable_()) {
+    // Helper functions for the workspaceChanged callback.
+    var getWorkspaceMetrics = function() {
+      var workspaceMetrics = Object.create(null);
+      var defaultMetrics = mainWorkspace.getMetrics();
+      var scale = mainWorkspace.scale;
+
+      // Get the view metrics in workspace units.
+      workspaceMetrics.viewLeft = defaultMetrics.viewLeft / scale;
+      workspaceMetrics.viewTop = defaultMetrics.viewTop / scale;
+      workspaceMetrics.viewRight =
+          (defaultMetrics.viewLeft + defaultMetrics.viewWidth) / scale;
+      workspaceMetrics.viewBottom =
+          (defaultMetrics.viewTop + defaultMetrics.viewHeight) / scale;
+
+      // Get the exact content metrics (in workspace units), even if the
+      // content is bounded.
+      if (mainWorkspace.isContentBounded_()) {
+        // Already in workspace units, no need to divide by scale.
+        var blocksBoundingBox = mainWorkspace.getBlocksBoundingBox();
+        workspaceMetrics.contentLeft = blocksBoundingBox.x;
+        workspaceMetrics.contentTop = blocksBoundingBox.y;
+        workspaceMetrics.contentRight =
+            blocksBoundingBox.x + blocksBoundingBox.width;
+        workspaceMetrics.contentBottom =
+            blocksBoundingBox.y + blocksBoundingBox.height;
+      } else {
+        workspaceMetrics.contentLeft = defaultMetrics.contentLeft / scale;
+        workspaceMetrics.contentTop = defaultMetrics.contentTop / scale;
+        workspaceMetrics.contentRight =
+            (defaultMetrics.contentLeft + defaultMetrics.contentWidth) / scale;
+        workspaceMetrics.contentBottom =
+            (defaultMetrics.contentTop + defaultMetrics.contentHeight) / scale;
+      }
+
+      return workspaceMetrics;
+    };
+    var getWorkspaceObjectMetrics = function(object) {
```

Comment:
This is a useful function to define directly on blocks and comments, instead of as a helper function that lives inside the inject code.  I'm pretty sure that we're doing exactly this in other places in the code as well, and we can transition those to use the new function as we find them.  It'll be minor code duplication, but it'll make this function easier to find and use (and annotate so it shows up in our documentation).

Reply:
Turns out it was already a function: getBoundingRectangle ([block](https://developers.google.com/blockly/reference/js/Blockly.BlockSvg#getBoundingRectangle), [comment](https://developers.google.com/blockly/reference/js/Blockly.WorkspaceCommentSvg#getBoundingRectangle)) the more you know ¯\\\_(ツ)_/¯